src/libPersistence/Retriever.cpp

```diff
@@ -50,26 +47,105 @@ bool Retriever::RetrieveTxBlocks(bool trimIncompletedBlocks) {
 
   unsigned int extra_txblocks = (lastBlockNum + 1) % NUM_FINAL_BLOCK_PER_POW;
 
-  if (trimIncompletedBlocks) {
-    // truncate the extra final blocks at last
-    for (unsigned int i = 0; i < extra_txblocks; ++i) {
-      BlockStorage::GetBlockStorage().DeleteTxBlock(lastBlockNum - i);
-      blocks.pop_back();
-    }
-  }
-
-  for (const auto& block : blocks) {
-    m_mediator.m_node->AddBlock(*block);
-  }
-
   /// Retrieve final block state delta from last DS epoch to
-  /// current TX epoch
+  /// current TX epoch and buffer the statedelta for each.
   for (const auto& block : blocks) {
     if (block->GetHeader().GetBlockNum() >= lastBlockNum + 1 - extra_txblocks) {
       bytes stateDelta;
       BlockStorage::GetBlockStorage().GetStateDelta(
           block->GetHeader().GetBlockNum(), stateDelta);
 
+      extraStateDeltas.push_back(stateDelta);
+    }
+  }
+
+  if ((lastBlockNum - extra_txblocks + 1) %
+          (INCRDB_DSNUMS_WITH_STATEDELTAS * NUM_FINAL_BLOCK_PER_POW) ==
+      0) {
+    // we must have latest state currently. so need not recreate states
+    LOG_GENERAL(INFO, "Current state is up-to-date until txblk :"
+                          << lastBlockNum - extra_txblocks);
+  } else {
+    // create states from last INCRDB_DSNUMS_WITH_STATEDELTAS *
```

Comment:
Hmmm sorry to say but actually this part of code is not needed. You can just keep DeserializeDelta from the last state checkpoint until the latest epoch.